Better RSpec parallelization, fixes #25

jenseng opened the pull request:
Once all groups have been assigned and a worker needs something to do, have it start running examples from one of the as-yet-unfinished groups. This way even if the last group is relatively large, the workers will divide and conquer, finishing around the same time. Hooks into ExampleGroup and friends and uses per-group example queues.

Unlike #26, we don't incur extra overhead from context-level hooks running on each example; they still just run once for each group. Performance overhead from the per-group example queues is negligible; sample_spec.rb's runtime is essentially unchanged. It might be worth making this the default behavior for RSpec once it's battle tested?

#### Other Notes

In order to isolate this to to the rspec* files, I had to do a little bit of refactoring/shuffling (primarily in the first commit). Open to suggestions if anything here seems too crazy :)

Stats are correctly aggregated when groups are split across runners, although they are fairly irrelevant now that we divide up the last groups.

This commit also updates test.sh to get rspec3*, and adds a separate spec file demonstrating the split groups performance.

/cc @bootstraponline @sbonebrake

bootstraponline commented:
:+1:

sbonebrake commented on lib/test_queue/runner/rspec.rb:
```diff
@@ -11,25 +11,129 @@
+        end
+      end
+
+      def has_own_examples?(group)
```
I'm having trouble with this terminology. Maybe "has_ descendant_examples_in_queue?" and "has_examples_in_queue?"

jenseng replied:
:+1: those are clearer names, I'll tweak this

sbonebrake commented:
First off, thanks a lot for taking the time to write all of this.  At a high level, the idea sounds great to me.  I'm still trying to wrap my head around some of the details of the implementation right now.  

I think we need still need the option to tag with :no_split or something similar for when example groups have examples that should never be executed at the same time due to external resource constraints (eg hard coded user, local db, etc).

I'm going to continue reviewing and trying to understand this next week.

jenseng commented:
Thanks for the feedback! I'll go ahead and add support for `:no_split`

One thing that's helpful in understanding this is to look at [`ExampleGroup.run`](https://github.com/rspec/rspec-core/blob/7425a5cab211483cf3660c4f2cb18a2da744b127/lib/rspec/core/example_group.rb#L536) and [`ExampleGroup.run_examples`](https://github.com/rspec/rspec-core/blob/7425a5cab211483cf3660c4f2cb18a2da744b127/lib/rspec/core/example_group.rb#L576) to see how the ExampleGroup extensions get used.

Basically, instead of iterating over the real `filtered_examples`, we instead get examples from the runner, one at a time. Typically this will be the full list, unless it's at the end of the run and another worker is helping us out. We have a queue for each group's (and sub-group's) immediate `filtered_examples`. That way we traverse the whole group hierarchy just once, running appropriate context-level hooks as we go.

There are probably ways to optimize this further, e.g. given 2 workers handling the following specs simultaneously:

```
Outer
  spec1
  spec2
  Inner1
    spec3
    spec4
  Inner2
    spec5
    spec6
```

you might see them divided up like so:

```
# worker 1
Outer
  spec1
  Inner1
    spec3
  Inner2
    spec5

# worker 2
Outer
  spec2
  Inner1
    spec4
  Inner2
    spec6

A better approach might assign out sub-groups as discrete units so that both workers don't have to run *all* the context hooks within the group. That said, it's really only an issue if you have expensive context-level hooks, and even then just in the groups that have simultaneous workers at the end, which will always be < N - 1 (where N is the number of workers). So it might not be worth the added complexity?

At any rate, let me know if you have any questions or other suggestions.

jenseng commented:
Just a couple tweaks... updated method names per your suggestion, and added support for `:no_split`. I also included some `:no_split` stuff in the spec so it's easy to see it working.

sbonebrake commented:
Ok, I think I understand it a bit better now.  Thanks for adding :no_split.  I did quite a few runs of my test suite (~300 tests, 150 workers) and I don't think it's running optimally.  The good news is I don't see any tests not running or running twice, but it seems we have a few workers that stop early and don't pick up tests that are later picked up by other workers.  I'm pretty sure that's not supposed to be possible with this execution strategy. I haven't figured out why this happens yet and I haven't been able to tweak your sample specs to reproduce it yet. I think I may need to fork and add debug statements each time a worker issues a request to the socket :/

jenseng commented:
@sbonebrake Thanks for testing it out, and let me know if I can assist. I haven't seen that issue in my usage, but I haven't exceeded 5 workers on our CI system yet (~500 specs)... workers always finish within seconds of each other. Are the total run times pretty different from one worker to the next?

While developing it, I found `TEST_QUEUE_VERBOSE` a tad misleading, since the entire output for each worker only displays once it completes (since each worker's stdout/stderr gets redirected to a file). If you run with `--format d` and comment out the stdout/stderr redirection in after_fork_internal, then you can see everything as it happens, all interleaved (terrible formatting though), and you should see the `Finished in x seconds` summary stuff all happening at the very end for all the runners.

jenseng commented:
Maybe some workers are getting broken pipes or something and quitting early? You could add some logging in [these](https://github.com/jenseng/test-queue/blob/b780652772612b44537cbe79ef2ea9704f503f11/lib/test_queue/iterator.rb#L30) [two](https://github.com/jenseng/test-queue/blob/b780652772612b44537cbe79ef2ea9704f503f11/lib/test_queue/iterator.rb#L68) rescues to see if that's the case.

sbonebrake commented:
Ok. I think I figured out the issue.  You actually mention the edge case in your comments, but at first I didn't fully realize the impact it has.  If there is a spec that has a medium-long setup block, it may be assigned to multiple workers.  These workers will ALL pick up the same ExampleGroup and begin execution when there is no example left to run.
Worst case is when there is an ExampleGroup with a long setup block that only contains a single example. Unless the setup block has a highly variable execution time, the 2nd+ worker to pick up the example group will never finish before the first worker to pick up the example group.

I have a small spec that exhibits the issue.
https://gist.github.com/sbonebrake/669b8ce9038ef480ee7d

At first I thought this would only happen towards the end of the run and although several workers may execute the same ExampleGroup with the long setup and single example, it would be fine because there is no "real work" left to do.  However, this isn't the case because there could be another ExampleGroup in the queue with a long setup and medium-long test cases such that the workers would would alternate picking up the ExampleGroup with long setup and "no real work" and the ExampleGroup that has "real work" left.  

Right now I don't know of a way of solving this unless you track the assignment of children examples from ExampleGroups. That way if an ExampleGroup with one example will only be given out once to a worker.  This would mean for each setup block, worst case is it is executed unnecessarily by 1 worker (because the initial ExampleGroup assignment finished it), rather than potentially all.

jenseng commented:
Good sleuthing! When I wrote that comment, I had envisioned this being an unlikely/insignificant edge case, but I could see this being pretty annoying in certain suites/setups.

If we track example assignments, couldn't there still be a race condition (but potentially smaller)? Like if the setup block is really slow, and many workers get assigned to the group before the initial worker even finishes the setup block and is assigned the one example?

I wonder if we could rethink the `has_descendant_examples_in_queue?` communication a bit and have that actually pop the next example in the group (immediate or descendant) if there is one. I think that way we could avoid having to do any extra/complex tracking on the runner side (though it would certainly complicate the ExampleGroup stuff a little bit). That should eliminate the race condition and ensure no workers run context hooks for groups they don't need to. What do you think?

I'll mull it over some more and take a crack at it tomorrow. Let me know if you have other thoughts or if you envision problems with this approach.

sbonebrake commented:
To clarify, I think the idea would be every time you assign the example
group you can assume it will execute 1 example and it would decrement the
number of examples in the group. Then it wouldn't assign groups with 0
children left. At most the setup would be executed the N times where N is
the number of examples. This definitely helps the case of the long setup
and the one quick test case, but I'm still thinking it wouldn't be optimal.
I'm starting to think about if we can do something with the stats file and
timing the setup as well. I'll keep brainstorming ideas.

On Wed, Dec 9, 2015, 8:45 PM Jon Jensen <notifications@github.com> wrote:

> Good sleuthing! When I wrote that comment, I had envisioned this being an
> unlikely/insignificant edge case, but I could see this being pretty
> annoying in certain suites/setups.
>
> If we track example assignments, couldn't there still be a race condition
> (but potentially smaller)? Like if the setup block is really slow, and many
> workers get assigned to the group before the initial worker even finishes
> the setup block and is assigned the one example?
>
> I wonder if we could rethink the has_descendant_examples_in_queue?
> communication a bit and have that actually pop the next example in the
> group (immediate or descendant) if there is one. I think that way we could
> avoid having to do any extra/complex tracking on the runner side (though it
> would certainly complicate the ExampleGroup stuff a little bit). That
> should eliminate the race condition and ensure no workers run context hooks
> for groups they don't need to. What do you think?
>
> I'll mull it over some more and take a crack at it tomorrow. Let me know
> if you have other thoughts or if you envision problems with this approach.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tmm1/test-queue/pull/31#issuecomment-163486690>.
>

jenseng commented:
Okay, I follow you. Like you say, it may still not be optimal, since N-1 workers might do unnecessary setup, if they all grab the group before the first worker finishes the setup, and that first worker then runs all the examples before they finish the setup.

The setup + stats idea is really interesting. It almost might be worth *only* tracking the setup time and not the spec time, and running the slowest groups first. That way if any groups do get parallelized at the very end, they are guaranteed to be the ones with the quickest setup, so it basically becomes a non-issue.

jenseng commented:
Found an unrelated problem that needs fixing in this PR ... was testing some fail-fast stuff, and workers seem to get stuck in an infinite loop with this PR (but not regular test-queue) if you set `RSpec.world.wants_to_quit = true` to make it bail.

jenseng commented:
No discussion? 😭

@tmm1 Objections to me submitting another PR along these lines (albeit cleaned up a bit, and with @sbonebrake's concerns addressed)?

Despite the lack of recent updates from me, we've actually been running [a variant](https://github.com/jenseng/test-queue/tree/betterer_rspec) of this PR for 6 months on canvas-lms w/ great success.... for our ridiculously large test suite (over 20k specs, 3k of them slooooow selenium), it gives us much better performance than either vanilla rspec-queue or the now-merged TEST_QUEUE_SPLIT_GROUPS feature.
